pkg/alertmanager/operator.go

```diff
@@ -469,6 +637,336 @@ func (c *Operator) sync(ctx context.Context, key string) error {
 	return nil
 }
 
+func (c *Operator) provisionAlertmanagerConfiguration(ctx context.Context, am *monitoringv1.Alertmanager) error {
+	// If no AlertmanagerConfig selectors are configured, the user wants to
+	// manage configuration themselves.
+	if am.Spec.AlertmanagerConfigSelector == nil {
+		level.Debug(c.logger).Log("msg", ", no AlertmanagerConfig selector specified, leaving configuration unmanaged", "alertmanager", am.Name, "namespace", am.Namespace)
+		return nil
+	}
+
+	secretName := defaultConfigSecretName(am.Name)
+	if am.Spec.ConfigSecret != "" {
```

Comment:
What should we use as a base if this is not set (it's currently optional)?

Reply:
If I understand the question correctly, then the current default secret name.

Reply:
Ah my bad, I get it now. If this is not set, and the "default secret" doesn't exist, then we handle that gracefully and still generate the configuration and mount the generated file. The change will essentially be that Alertmanager pods will be scheduled even if users don't configure routes (previously the base secret had to be present).

Reply:
@brancz thanks for replying! :) 

Currently (AIUI), if the field is not set, we assume that there's a secret there with the default name. Since we don't read the secret (just mount it as a volume, let k8s handle if it's missing or not).

Now, given that we're making a new Secret, if there's no Secret we generate a skeleton base, so the config can be generated from the AlertmanagerConfig CRs? I'll try that. I think that means that some error conditions here can go away: we don't error if there's no secret, and we could also not error if there _is_ a config secret, but it doesn't contain a base route (we could just add one with a null receiver (and add that receiver if it's missing)? If a base secret does get added (or the missing route gets added to the existing one), then we'll be regenerating the mounted secret anyway, so everything should work.

Does that sound reasonable to you @simonpasquier?

Reply:
I've changed it to something along the lines of what I was thinking about, let me know what you think!

I wonder if a validating webhook for `Alertmanager` resources would be a good idea? i.e. reject submission of a CR if the `configSecret` is set to a missing secret, or to a secret that doesn't have the expected field, or if that field can't be parsed to a valid Alertmanager config. That way people would know immediately that something is wrong, rather than the operator either failing or doing something that they didn't intend.

Reply:
agree with your approach Gerard. And a validating webhook would be a great addition!

Reply:
I started looking at adding to the existing admission webhook stuff, and how to add more to it, but then I stopped because it doesn't seem to be used by default. I've got some wip changes to it locally, but I think those might be better as a separate change rather than adding more to this one. Let me know if you think it should be in this PR and I can finish it and add here. :+1: